### PR TITLE
python - add pointwisemult and axpy

### PR DESCRIPTION
**Pull request by jeremylt**

### CODE REVIEW

**Comment by jeremylt** [comment]
@DiffeoInvariant GitHub isn't letting me add you to review, but here's the PR.

**Comment by DiffeoInvariant** [comment]
@jeremylt I can't review because I don't have write access to the libCEED repo, but this looks good to me

**Comment by jeremylt** [comment]
CUDA and HIP impls added

File: backends/cuda/ceed-cuda-vector.c
Comment (jeremylt) [inline backends/cuda/ceed-cuda-vector.c:443]: I added this when I first added the device versions of the vector reciprocal function. (It is actually buggy as it is in main; needs line 443, which I added in previous commit)

I'm dropping this sync in this commit - we generally have a philosophy of not forcing a sync unless we need it, so I see no reason to maintain `CEED_*_BOTH_SYNC` in this case. If the data needs to be on the host, it can be pulled down there when needed.

Thoughts @YohannDudouit @nbeams?

**Review by DiffeoInvariant — APPROVED** [review]
This looks good to me! I've never written CUDA code myself so I probably shouldn't sign off on that, but the Python and other small changes look good to me

File: backends/cuda/ceed-cuda-vector.c
Comment (nbeams) [reply backends/cuda/ceed-cuda-vector.c:443]: Makes sense to me.

File: backends/hip/kernels/hip-vector.hip.cpp
Comment (nbeams) [inline backends/hip/kernels/hip-vector.hip.cpp:89]: This line reminds me that I probably shouldn't have left all the vector blocksizes at 512 for HIP after we discovered the issue with the compiler assuming 256 threads/block.  And now we don't match either the 4.0 assumption (256) or 4.1 assumption (1024), so perhaps this is something I should look into more...though with such simple kernels, it's unlikely to cause failures (we're not exactly taxing the register or shared memory usage here), I guess there's still a chance performance could change.  I did check the detailed compiler output from the `--save-temps` flag (something Scott told me about) and the result with/without adding `__launch_bounds(512)` is identical.

So I guess I'd be fine with leaving it for now, but we could also switch to 256 or add `__launch_bounds__` to all the kernels in this file for completeness.  (Since 4.1 assumes 1024, there will no longer be the potential for issues with errors like were seen previously.)   Or we could leave this for a future PR since we might want to add launch bounds other places after the change in 4.1; I've currently been investigating this quite a bit with the hipMAGMA backend because the results with/without launch bounds even with 4.0 are not what I expected (haven't been able to try 4.1 yet).

**Review by nbeams — APPROVED** [review]
Looks good to me.  

One thing I happened to notice that isn't part of this PR (but this PR may be a good place to fix it) is the typo "poinwise" in the description comments at the top of `t122-vector.c`.

File: backends/hip/kernels/hip-vector.hip.cpp
Comment (jedbrown) [reply backends/hip/kernels/hip-vector.hip.cpp:89]: Should I upgrade Noether to 4.1? I left it at 4.0 because Scott said there were known issues and I didn't have enough context to know whether those were likely to impact us.

Comment (nbeams) [reply backends/hip/kernels/hip-vector.hip.cpp:89]: I'm not sure -- I was under the impression that the issues would maybe affect performance more than cause failures -- were you thinking the same?  Maybe we can ask Scott for more clarification.  I don't know what all the users of Noether need, either... I mostly just test/debug (and am very grateful for the access on days like today, when the machine I use here at ICL has been busy all day with others' tests) but obviously I don't know what everyone else uses Noether for.

Comment (jedbrown) [reply backends/hip/kernels/hip-vector.hip.cpp:89]: If it's pure performance, then not a big deal. We sometimes test performance, but don't put any particular confidence on it.

**Comment by jeremylt** [comment]
I fixed the typo. I'll merge once CI finishes.

**Review by YohannDudouit — APPROVED** [review]

**Pull request merged**
